provisioning/device/src/Transports/Amqp/AmqpClientConnection.cs

```diff
@@ -34,121 +38,149 @@ internal AmqpClientConnection(Uri uri, AmqpSettings amqpSettings)
+                                $"Sent Protocol Header: {_sentHeader} operationPending: {operationPending} completedSynchronously: {args.CompletedSynchronously}");
+
+                        if (!operationPending)
+                        {
+                            args.CompletedCallback(args);
+                        }
+
+                        _transport = await _tcs.Task.ConfigureAwait(false);
+                        await _transport.OpenAsync(cancellationToken).ConfigureAwait(false);
                     }
-
-                    _transport = await _tcs.Task.ConfigureAwait(false);
-                    await _transport.OpenAsync(cancellationToken).ConfigureAwait(false);
                 }
+                else
+                {
+                    var tcpInitiator = new AmqpTransportInitiator(_amqpSettings, TransportSettings);
+                    _transport = await tcpInitiator.ConnectAsync(cancellationToken).ConfigureAwait(false);
+                }
+
+                AmqpConnection = new AmqpConnection(_transport, _amqpSettings, AmqpConnectionSettings);
+                AmqpConnection.Closed += OnConnectionClosed;
+                await AmqpConnection.OpenAsync(cancellationToken).ConfigureAwait(false);
+                IsConnectionClosed = false;
             }
-            else
+            finally
             {
-                var tcpInitiator = new AmqpTransportInitiator(_amqpSettings, TransportSettings);
-                _transport = await tcpInitiator.ConnectAsync(cancellationToken).ConfigureAwait(false);
-            }
+                _connectionSemaphore.Release();
 
-            AmqpConnection = new AmqpConnection(_transport, _amqpSettings, AmqpConnectionSettings);
-            AmqpConnection.Closed += OnConnectionClosed;
-            await AmqpConnection.OpenAsync(cancellationToken).ConfigureAwait(false);
-            IsConnectionClosed = false;
+                if (Logging.IsEnabled)
+                    Logging.Exit(this, $"{nameof(AmqpClientConnection)}.{nameof(OpenAsync)}");
+            }
         }
 
-        public async Task CloseAsync(CancellationToken cancellationToken)
+        internal async Task CloseAsync(CancellationToken cancellationToken)
         {
-            AmqpConnection connection = AmqpConnection;
-            if (connection != null)
+            if (AmqpConnection == null)
             {
-                await connection.CloseAsync(cancellationToken).ConfigureAwait(false);
+                return;
             }
-        }
 
-        public void Close()
-        {
-            AmqpConnection connection = AmqpConnection;
-            if (connection != null)
+            await _connectionSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+            try
             {
-                connection.Close();
+                await AmqpConnection.CloseAsync(cancellationToken).ConfigureAwait(false);
+                AmqpConnection = null;
```

Comment:
Was this ever being set to null anywhere else?

Reply:
Nope.